### PR TITLE
Implement subroutines for yul functions.

### PR DESCRIPTION
**Pull request by chriseth**
Problems with block deduplicator:
 - [ ] it might unify a beginsub tag with a regular tag
 - [ ] it might unify tags and thus turn a sub-internal jump into a cross-sub-jump

It might make sense to run the block deduplicator on subroutines in isolation.

### CODE REVIEW

File: libevmasm/Assembly.cpp [hunk 1]
```diff
@@ -173,7 +177,7 @@ void Assembly::assemblyStream(ostream& _out, string const& _prefix, StringMap co
 
 	if (!m_data.empty() || !m_subs.empty())
 	{
-		_out << _prefix << "stop" << endl;
+		_out << _prefix << "beginsub" << endl;
```
Comment (axic) [inline libevmasm/Assembly.cpp:180]: This has to be versioned (`evmVersion >= berlin ? beginsub : stop`)

File: libevmasm/Assembly.cpp [hunk 2]
```diff
@@ -702,8 +713,8 @@ LinkerObject const& Assembly::assemble() const
 
 
 	if (!m_subs.empty() || !m_data.empty() || !m_auxiliaryData.empty())
-		// Append an INVALID here to help tests find miscompilation.
-		ret.bytecode.push_back(uint8_t(Instruction::INVALID));
+		// Append a BEGINSUB here to help tests find miscompilation.
+		ret.bytecode.push_back(uint8_t(Instruction::BEGINSUB));
```
Comment (axic) [inline libevmasm/Assembly.cpp:717]: This should be versioned (evmVersion >= berlin ? beginsub : invalid)

Comment (chriseth) [reply libevmasm/Assembly.cpp:717]: I thought so, too, but `BEGINSUB` is an invalid opcode prior to berlin...

Comment (axic) [reply libevmasm/Assembly.cpp:717]: That is true, but one of the outstanding question is whether any deployed contract contains a BEGINSUB opcode between a JUMP/JUMPDEST which would render that JUMP broken.

Iff this PR were to be merged **and** people would deploy contracts with this change, that would increase the number of contracts to be analyzed.

In short, we'd be nicer citizens keeping `INVALID` there prior to Berlin :)

**Comment by chriseth** [comment]
@axic requesting to close this.

**Pull request closed without merging**